File: src/adapters/qwen/install_qwen_integration.py
"""
Qwen CLI Integration Installer

Sets up basic integration for Qwen CLI with cross-CLI collaboration capabilities
"""

import os
import json
import sys
import shutil
import argparse
from pathlib import Path
from datetime import datetime

# Qwen CLI config paths
QWEN_CONFIG_DIR = os.path.expanduser("~/.qwen")
QWEN_CONFIG_FILE = os.path.join(QWEN_CONFIG_DIR, "config.json")

def create_qwen_config_directory():
    """Create Qwen config directory"""
    os.makedirs(QWEN_CONFIG_DIR, exist_ok=True)
    print("[OK] Created Qwen config directory: {}".format(QWEN_CONFIG_DIR))

def install_qwen_config():
    """Install basic Qwen CLI config for cross-CLI integration"""
    # Read existing config
    existing_config = {}
    if os.path.exists(QWEN_CONFIG_FILE):
        try:
            with open(QWEN_CONFIG_FILE, 'r', encoding='utf-8') as f:
                existing_config = json.load(f)
        except Exception as e:
            print("Warning: Failed to read existing config: {}".format(e))
            existing_config = {}

    # Define cross-CLI integration config
    cross_cli_config = {
        "cross_cli_enabled": True,
        "supported_clis": ["claude", "gemini", "iflow", "qodercli", "codebuddy", "copilot", "codex"],
        "auto_detect": True,
        "timeout": 30,
        "collaboration_mode": "active"
    }

    # Merge configs
    merged_config = existing_config.copy()
    merged_config.update(cross_cli_config)

    # Write config file
    try:
        with open(QWEN_CONFIG_FILE, 'w', encoding='utf-8') as f:
            json.dump(merged_config, f, indent=2, ensure_ascii=False)

        print("[OK] Qwen config installed: {}".format(QWEN_CONFIG_FILE))
        return True
    except Exception as e:
        print("Error: Failed to install Qwen config: {}".format(e))
        return False

def copy_adapter_files():
    """Copy adapter files to Qwen config directory"""
    try:
        # Get current script directory
        current_dir = Path(__file__).parent
        
        # Look for any Python adapter files to copy
        adapter_files = list(current_dir.glob("*.py"))
        adapter_files = [f for f in adapter_files if f.name != Path(__file__).name]
        
        if not adapter_files:
            print("Note: No adapter files found to copy")
            return True
            
        # Copy adapter files
        for src_file in adapter_files:
            dst_file = os.path.join(QWEN_CONFIG_DIR, src_file.name)
            shutil.copy2(src_file, dst_file)
            print("[OK] Copied adapter file: {}".format(src_file.name))
            
        return True
    except Exception as e:
        print("Warning: Failed to copy adapter files: {}".format(e))
        return True  # Don't fail the entire installation for this

def verify_installation():
    """Verify installation success"""
    print("\nVerifying Qwen CLI integration installation...")

    # Check config file
    if not os.path.exists(QWEN_CONFIG_FILE):
        print("Warning: Config file does not exist: {}".format(QWEN_CONFIG_FILE))
        return True  # Not critical for basic functionality

    # Check config file content
    try:
        with open(QWEN_CONFIG_FILE, 'r', encoding='utf-8') as f:
            config = json.load(f)
            
        if config.get('cross_cli_enabled'):
            print("[OK] Cross-CLI integration enabled")
        else:
            print("Note: Cross-CLI integration not enabled")
            
        print("[OK] Qwen config file verified")
        return True
    except Exception as e:
        print("Warning: Failed to verify config file: {}".format(e))
        return True  # Not critical for basic functionality

def uninstall_integration():
    """Uninstall Qwen integration"""
    try:
        # Remove cross-CLI config from config file
        if os.path.exists(QWEN_CONFIG_FILE):
            with open(QWEN_CONFIG_FILE, 'r', encoding='utf-8') as f:
                config = json.load(f)

            # Remove cross-CLI settings
            config.pop('cross_cli_enabled', None)
            config.pop('supported_clis', None)
            config.pop('auto_detect', None)
            config.pop('collaboration_mode', None)

            # Save updated config
            with open(QWEN_CONFIG_FILE, 'w', encoding='utf-8') as f:
                json.dump(config, f, indent=2, ensure_ascii=False)

            print("[OK] Removed cross-CLI settings from Qwen config")

        print("[OK] Qwen CLI cross-CLI integration uninstalled")
        return True
    except Exception as e:
        print("Error: Uninstall failed: {}".format(e))
        return False

def main():
    """Main function"""
    parser = argparse.ArgumentParser(description="Qwen CLI Integration Installer")
    parser.add_argument("--verify", action="store_true", help="Verify installation")
    parser.add_argument("--uninstall", action="store_true", help="Uninstall integration")
    args = parser.parse_args()

    print("Qwen CLI Integration Installer")
    print("=" * 40)

    if args.uninstall:
        return uninstall_integration()
    elif args.verify:
        return verify_installation()
    else:
        # Execute installation
        print("Step 1. Creating config directory...")
        create_qwen_config_directory()

        print("\nStep 2. Installing config...")
        config_success = install_qwen_config()

        print("\nStep 3. Copying adapter files...")
        adapter_success = copy_adapter_files()

        print("\nStep 4. Verifying installation...")
        verification_success = verify_installation()

        overall_success = config_success and adapter_success and verification_success
        if overall_success:
            print("\n[SUCCESS] Qwen CLI integration installed successfully!")
        else:
            print("\n[WARNING] Installation completed with warnings")

        return overall_success

if __name__ == "__main__":
    success = main()
    sys.exit(0 if success else 1)

File: src/adapters/qwen/install_qwen_integration.js
#!/usr/bin/env node

/**
 * Qwen CLI Integration Installer - JavaScript Version
 * Sets up basic integration for Qwen CLI with cross-CLI collaboration capabilities
 */

const fs = require('fs').promises;
const path = require('path');
const os = require('os');

// Qwen CLI config paths
const QWEN_CONFIG_DIR = path.join(os.homedir(), '.qwen');
const QWEN_CONFIG_FILE = path.join(QWEN_CONFIG_DIR, 'config.json');
const QWEN_HOOKS_FILE = path.join(QWEN_CONFIG_DIR, 'hooks.json');

class QwenInstaller {
  constructor() {
    this.toolName = 'qwen';
    this.configDir = QWEN_CONFIG_DIR;
    this.configFile = QWEN_CONFIG_FILE;
    this.hooksFile = QWEN_HOOKS_FILE;
  }

  async createConfigDirectory() {
    await fs.mkdir(this.configDir, { recursive: true });
    console.log(`[OK] Created Qwen config directory: ${this.configDir}`);
  }

  async installConfig() {
    // Read existing config
    let existingConfig = {};
    try {
      const content = await fs.readFile(this.configFile, 'utf-8');
      existingConfig = JSON.parse(content);
    } catch (e) {
      console.log(`Warning: Failed to read existing config: ${e.message}`);
      existingConfig = {};
    }

    // Define cross-CLI integration config
    const crossCliConfig = {
      cross_cli_enabled: true,
      supported_clis: ['claude', 'gemini', 'qwen', 'iflow', 'qodercli', 'codebuddy', 'copilot', 'codex'],
      auto_detect: true,
      timeout: 30,
      collaboration_mode: 'active',
      qwen_oauth_integration: true
    };

    // Merge configs
    const mergedConfig = { ...existingConfig, ...crossCliConfig };

    // Write config file
    await fs.writeFile(this.configFile, JSON.stringify(mergedConfig, null, 2));
    console.log(`[OK] Qwen config installed: ${this.configFile}`);
    return true;
  }

  async installHooks() {
    // Read existing hooks config
    let existingHooks = {};
    try {
      const content = await fs.readFile(this.hooksFile, 'utf-8');
      existingHooks = JSON.parse(content);
    } catch (e) {
      console.log(`Warning: Failed to read existing hooks config: ${e.message}`);
      existingHooks = {};
    }

    // Define cross-CLI integration hooks
    const crossCliHooks = {
      cross_cli_adapter: {
        enabled: true,
        supported_tools: ['claude', 'gemini', 'qwen', 'iflow', 'qodercli', 'codebuddy', 'copilot', 'codex'],
        trigger_patterns: [
          'use\\s+(\\w+)\\s+to\\s+(.+)$',
          'call\\s+(\\w+)\\s+(.+)$',
          'ask\\s+(\\w+)\\s+(.+)$',
          'stigmergy\\s+(\\w+)\\s+(.+)$'
        ]
      }
    };

    // Merge hooks configs
    const mergedHooks = { ...existingHooks, ...crossCliHooks };

    // Write hooks config file
    await fs.writeFile(this.hooksFile, JSON.stringify(mergedHooks, null, 2));
    console.log(`[OK] Qwen hooks installed: ${this.hooksFile}`);
    return true;
  }

  async copyAdapterFiles() {
    const currentDir = __dirname;
    const targetDir = path.join(this.configDir, 'hooks');
    
    try {
      // Look for adapter files to copy
      const adapterFiles = await fs.readdir(currentDir);
      const filesToCopy = adapterFiles.filter(file => 
        file.endsWith('.js') || file.endsWith('.json') || file.endsWith('.py') || file === '__init__.py'
      );

      for (const file of filesToCopy) {
        const srcFile = path.join(currentDir, file);
        const dstFile = path.join(targetDir, file);
        await fs.copyFile(srcFile, dstFile);
        console.log(`[OK] Copied adapter file: ${file}`);
      }

      // Create global Cross-CLI documentation
      await this.createGlobalCrossCliDocumentation(this.configDir);
      
      return true;
    } catch (e) {
      console.log(`Warning: Failed to copy adapter files: ${e.message}`);
      return true;
    }
  }

  async createGlobalCrossCliDocumentation(configDir) {
    try {
      const docContent = `# Qwen CLI Cross-CLI Communication Guide

## Cross-CLI Communication
When you need to call other AI tools, please use the following prompt <<<Execute the following command in shell: stigmergy distcli "Prompt">>>

Examples:
- Run in shell: stigmergy claude "write a Python function"
- Run in shell: stigmergy gemini "translate this text"
- Run in shell: stigmergy qwen "analyze this code"

Available tools: claude, gemini, qwen, iflow, qodercli, codebuddy, copilot, codex

---
*This document is automatically generated and maintained by Stigmergy CLI*
`;
      
      const docPath = path.join(configDir, 'CROSS_CLI_GUIDE.md');
      await fs.writeFile(docPath, docContent);
      console.log(`[OK] Created Cross-CLI Communication Guide: ${docPath}`);
      
      // Append to qwen.md if it exists
      const qwenMdPath = path.join(os.homedir(), '.qwen', 'qwen.md');
      try {
        await fs.access(qwenMdPath);
        const crossCliContent = `

## Cross-CLI Communication
When you need to call other AI tools, please use the following prompt <<<Execute the following command in shell: stigmergy distcli "Prompt">>>

Examples:
- Run in shell: stigmergy claude "write a Python function"
- Run in shell: stigmergy gemini "translate this text"
- Run in shell: stigmergy qwen "analyze this code"

Available tools: claude, gemini, qwen, iflow, qodercli, codebuddy, copilot, codex
`;
        await fs.appendFile(qwenMdPath, crossCliContent);
        console.log('[OK] 在QWEN.md末尾追加Cross-CLI通信提示');
      } catch (e) {
        // File doesn't exist, that's ok
      }
      
      return true;
    } catch (e) {
      console.log(`Warning: Failed to create Cross-CLI Communication Guide: ${e.message}`);
      return false;
    }
  }

  async verifyInstallation() {
    console.log('\n验证Qwen CLI集成安装...');

    // Check config directory
    if (!await fs.access(this.configDir).then(() => true).catch(() => false)) {
      console.log(`Warning: Config directory does not exist: ${this.configDir}`);
      return true;
    }

    // Check config file content
    try {
      const content = await fs.readFile(this.configFile, 'utf-8');
      const config = JSON.parse(content);
      
      if (config.cross_cli_enabled) {
        console.log('[OK] Cross-CLI integration enabled');
      } else {
        console.log('Note: Cross-CLI integration not enabled');
      }
      
      console.log('[OK] Qwen config file verified');
      return true;
    } catch (e) {
      console.log(`Warning: Failed to verify config file: ${e.message}`);
      return true;
    }
  }

  async uninstallIntegration() {
    try {
      // Remove cross-CLI config from config file
      if (await fs.access(this.configFile).then(() => true).catch(() => false)) {
        const content = await fs.readFile(this.configFile, 'utf-8');
        const config = JSON.parse(content);

        // Remove cross-CLI settings
        delete config.cross_cli_enabled;
        delete config.supported_clis;
        delete config.auto_detect;
        delete config.collaboration_mode;
        delete config.qwen_oauth_integration;

        // Save updated config
        await fs.writeFile(this.configFile, JSON.stringify(config, null, 2));
        console.log('[OK] Removed cross-CLI settings from Qwen config');
      }

      // Remove hooks config file
      if (await fs.access(this.hooksFile).then(() => true).catch(() => false)) {
        await fs.unlink(this.hooksFile);
        console.log('[OK] Removed Qwen hooks config file');
      }

      console.log('[OK] Qwen CLI cross-CLI integration uninstalled');
      return true;
    } catch (e) {
      console.log(`Error: Uninstall failed: ${e.message}`);
      return false;
    }
  }

  async install() {
    console.log('Qwen CLI Integration Installer');
    console.log('==========================================');

    // Execute installation
    console.log('Step 1. 创建配置目录...');
    await this.createConfigDirectory();

    console.log('\nStep 2. 安装配置...');
    const configSuccess = await this.installConfig();

    console.log('\nStep 3. 安装钩子...');
    const hooksSuccess = await this.installHooks();

    console.log('\nStep 4. 复制适配器文件...');
    const adapterSuccess = await this.copyAdapterFiles();

    console.log('\nStep 5. 验证安装...');
    const verificationSuccess = await this.verifyInstallation();

    const overallSuccess = configSuccess && hooksSuccess && adapterSuccess && verificationSuccess;
    if (overallSuccess) {
      console.log('\n[SUCCESS] Qwen CLI integration installed successfully!');
    } else {
      console.log('\n[WARNING] Installation completed with warnings');
    }

    return overallSuccess;
  }
}

// Main execution
if (require.main === module) {
  const installer = new QwenInstaller();
  
  const args = process.argv.slice(2);
  const command = args[0];

  switch (command) {
  case '--verify':
    installer.verifyInstallation().then(success => process.exit(success ? 0 : 1));
    break;
  case '--uninstall':
    installer.uninstallIntegration().then(success => process.exit(success ? 0 : 1));
    break;
  default:
    installer.install().then(success => process.exit(success ? 0 : 1));
    break;
  }
}

module.exports = QwenInstaller;